### PR TITLE
Rewritten for preview 4 to use the new resource model

### LINKED ISSUE #336: Use Application Insights for .NET Aspire telemetry Doc not clear
**Issue description**

Currently, the Use Application Insights for .NET Aspire telemetry doc is unclear on what exactly needs to be done to configure an Aspire app with Applicaiton Insights and deploy it to Azure Container Apps with the Azure Developer CLI. In the doc, it isn't clear how a new Application Insights resource is created as well as how you map the connection string from that resource and add it to the environment of ACA. 

Today if you add the code referenced to the 2 projects and run `azd up` or `azd deploy` it will create environment variables in ACA but they are empty. What really needs to happen is that secrets need to be created on ACA side and azd needs to read those in and attach them to the Revision in ACA. 

**Target framework**

- [x] .NET Core
- [ ] .NET Framework
- [ ] .NET Standard

<details>
<summary><strong>dotnet --info output</strong> or <strong>About VS info</strong></summary>

```console
dotnet --info
.NET SDK:
 Version:           8.0.200-preview.23624.5
 Commit:            8065b9770c
 Workload version:  8.0.200-manifests.ba313bcd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22635
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.200-preview.23624.5\

.NET workloads installed:
 Workload version: 8.0.200-manifests.ba313bcd
 [aspire]
   Installation Source: SDK 8.0.200-preview.23624, VS 17.9.34518.117
   Manifest Version:    8.0.0-preview.2.23619.3/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.0.0-preview.2.23619.3\WorkloadManifest.json
   Install Type:              Msi

 [maccatalyst]
   Installation Source: VS 17.9.34518.117
   Manifest Version:    17.2.8004/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maccatalyst\17.2.8004\WorkloadManifest.json
   Install Type:              Msi

 [ios]
   Installation Source: VS 17.9.34518.117
   Manifest Version:    17.2.8004/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.ios\17.2.8004\WorkloadManifest.json
   Install Type:              Msi

 [android]
   Installation Source: VS 17.9.34518.117
   Manifest Version:    34.0.52/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.android\34.0.52\WorkloadManifest.json
   Install Type:              Msi

 [maui-windows]
   Installation Source: VS 17.9.34518.117
   Manifest Version:    8.0.3/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maui\8.0.3\WorkloadManifest.json
   Install Type:              Msi


Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71

.NET SDKs installed:
  6.0.418 [C:\Program Files\dotnet\sdk]
  7.0.312 [C:\Program Files\dotnet\sdk]
  7.0.405 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.2.23502.2 [C:\Program Files\dotnet\sdk]
  8.0.200-preview.23624.5 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.2.23480.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.2.23479.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.2.23479.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
</details>

### LINKED ISSUE #431: AppInsights article needs to be refreshed for P4.
### Type of issue

Outdated article

### Description

We need some significant changes here for P4. Check in with @BrennanConroy and @samsp-msft for final details.

### Page URL

https://learn.microsoft.com/en-us/dotnet/aspire/deployment/azure/application-insights

### Content source URL

https://github.com/dotnet/docs-aspire/blob/main/docs/deployment/azure/application-insights.md

### Document Version Independent Id

fee898a4-8244-7139-151f-38f7c49e9f45

### Article author

@IEvangelist

### PR DESCRIPTION
**Pull request by samsp-msft**
## Summary

Describe your changes here.

- Fixes #431
- Fixes #336

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/deployment/azure/application-insights.md](https://github.com/dotnet/docs-aspire/blob/07ce9deab203c24e563baca0ae7f20442ca89918/docs/deployment/azure/application-insights.md) | [Use Application Insights for .NET Aspire telemetry](https://review.learn.microsoft.com/en-us/dotnet/aspire/deployment/azure/application-insights?branch=pr-en-us-454) |

### CODE REVIEW

**Comment by samsp-msft** [comment]
@IEvangelist - is it right to pull this to preview 4 rather than main, as its P4 content.

**Comment by IEvangelist** [comment]
> @IEvangelist - is it right to pull this to preview 4 rather than main, as its P4 content.

Yeah, I'll update this PR to target that branch instead.

**Comment by IEvangelist** [comment]
> @IEvangelist - is it right to pull this to preview 4 rather than main, as its P4 content.

Hi @samsp-msft - absolutely, that's perfect. Thank you

**Comment by IEvangelist** [comment]
Hey, @samsp-msft did you see #451?

File: docs/deployment/azure/application-insights.md [hunk 1]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
+
+To use automatic provisioning you will need to specify a dependency in the AppHost project, and reference it in each project/resource that needs to send telemetry to Application Insights. The steps include:
+
+- Adding a Nuget package reference to `Aspire.Hosting.Azure.Provisioning` in the AppHost project. This will also implicitly include the `Aspire.Hosting.Azure` package.
+
+- Add [app secrets](/aspnet/core/security/app-secrets) to tell it where to create the Application Insights Resource. If using Visual Studio, right click on the AppHost project and choose *Manage User Secrets* to create and open the secrets file. Add the following keys, replacing the values as applicable:
+
+``` json
+{
+  "Azure:Location": "WestUS",
```
Comment (BrennanConroy) [inline docs/deployment/azure/application-insights.md:32]: nit:
```
"Azure": {
  "Location": "West US",
  "SubscriptionId": "__Replace with your subscription GUID__"
}
```

Comment (samsp-msft) [reply docs/deployment/azure/application-insights.md:32]: Is this the id or the name of the region? We should be clear here.

Comment (BrennanConroy) [reply docs/deployment/azure/application-insights.md:32]: Idk, I just used "West US 2" when I was trying earlier.

File: docs/deployment/azure/application-insights.md [hunk 2]
```diff
@@ -73,10 +164,12 @@ private static IHostApplicationBuilder AddOpenTelemetryExporters(
 
     // Uncomment the following lines to enable the Azure Monitor exporter 
     // (requires the Azure.Monitor.OpenTelemetry.Exporter package)
```
Comment (BrennanConroy) [inline docs/deployment/azure/application-insights.md:166]:
```suggestion
    // (requires the Azure.Monitor.OpenTelemetry.AspNetCore package)
```

File: docs/deployment/azure/application-insights.md (hunk 1 above)
Comment (samsp-msft) [reply docs/deployment/azure/application-insights.md:32]: who would know @davidfowl?

Comment (IEvangelist) [reply docs/deployment/azure/application-insights.md:32]: The location and subscription id pair are always:

- `Location`: A value from a result in this table, `az account list-locations -o table`. Best to use the `Name`, but I believe `DisplayName` works too.
- `SubscriptionId`: GUID for your sub.

File: docs/deployment/azure/application-insights.md [hunk 3]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
 ```json
 {
-  "APPLICATIONINSIGHTS_CONNECTION_STRING": "InstrumentationKey=12345678-abcd-1234-abcd-1234abcd5678;IngestionEndpoint=https://westus3-1.in.applicationinsights.azure.com"
+  "ConnectionStrings": {
+    "myInsightsResource": "InstrumentationKey=12345678-abcd-1234-abcd-1234abcd5678;IngestionEndpoint=https://westus3-1.in.applicationinsights.azure.com"
+  }
 }
 ```
 
+> Note: The keyname specified in the AppHost code needs to match a key inside the `ConnectionStrings` section in the settings file.
```
Comment (IEvangelist) [inline docs/deployment/azure/application-insights.md:109]:
```suggestion
> [!NOTE]
> The `name` specified in the app host code needs to match a key inside the `ConnectionStrings` section in the settings file.
```

File: docs/deployment/azure/application-insights.md [hunk 4]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
+If you wish to use a different mechanism at developemnt from deployment, you can use a combination of Application Insights and Connection String resources, with a conditional. For example the following code uses an automatically provisioned resource at development time, and a pre-supplied connection at deployment time.
+
+```
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddAzureProvisioning();
+
+var insights = builder.ExecutionContext.IsPublishMode ?
+    builder.AddConnectionString("ai", "APPLICATIONINSIGHTS_CONNECTION_STRING") :
+    builder.AddAzureApplicationInsights("ai");
```
Comment (IEvangelist) [inline docs/deployment/azure/application-insights.md:126]:
```suggestion
var insights = builder.ExecutionContext.IsPublishMode
    ? builder.AddConnectionString("ai", "APPLICATIONINSIGHTS_CONNECTION_STRING")
    : builder.AddAzureApplicationInsights("ai");
```

File: docs/deployment/azure/application-insights.md [hunk 5]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
 
+> Note: The keyname specified in the AppHost code needs to match a key inside the `ConnectionStrings` section in the settings file.
+
+#### Resource usage during deployment
+
+When [deploying an Aspire application with Azure Developer CLI (AZD)](./aca-deployment-azd-in-depth.md), it will recognize the connection string resource and prompt for a value. This enables a different resource to be used for the deployment from the value used for local development.
+
+### Mixed deployment
+
+If you wish to use a different mechanism at developemnt from deployment, you can use a combination of Application Insights and Connection String resources, with a conditional. For example the following code uses an automatically provisioned resource at development time, and a pre-supplied connection at deployment time.
```
Comment (IEvangelist) [inline docs/deployment/azure/application-insights.md:117]:
```suggestion
If you wish to use a different deployment mechanism per execution context, use the appropriate API conditionally. For example, the following code uses an automatically provisioned resource at development time, and a pre-supplied connection at deployment time.
```

File: docs/deployment/azure/application-insights.md [hunk 6]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
+    .WithReference(insights);
+
+builder.AddProject<Projects.aspire_app11_Web>("webfrontend")
+    .WithReference(apiService)
+    .WithReference(insights);
+
+builder.Build().Run();
+```
+
+> Note: With the above you will need to supply the `Azure:Location` and `Azure:Subscription` information in app secrets for development time usage, and will be prompted for the connection string by AZD at deployment time.
```
Comment (IEvangelist) [inline docs/deployment/azure/application-insights.md:138]: Please confirm my suggested change here, it's `Azure:SubscriptionId`, right? And not `Azure:Subscription`.

```suggestion
> [!TIP]
> The preceding code requires you to supply the `Azure:Location` and `Azure:SubscriptionId` information in app secrets for development time usage, and will be prompted for the connection string by AZD at deployment time.
```

**Review by IEvangelist — APPROVED** [review]

File: docs/deployment/azure/application-insights.md [hunk 7]
```diff
@@ -9,60 +9,151 @@ ms.topic: how-to
...
+When you debug the application, Aspire will create the Application Insights resource on the first time that it sees the resource. The Application Insights Resource will feed the connection information to each project that references it. This can be seen as an `APPLICATIONINSIGHTS_CONNECTION_STRING` environment variable in the details view for the resource in the Aspire dashboard.
+
+:::image type="content" source="../media/dashboard-with-connection-string.png" lightbox="../media/dashboard-with-connection-string.png" alt-text="Connection string environment variables in the Aspire dashboard":::
+
+The settings for the created Application Insights resource will be persisted in the app secrets file.
+
+### Provisioning Application insights at deployment time
+
+Using the Application Insights resource above will cause an Application Insights resource to be provisioned when the application is deployed using the Azure Developer CLI (AZD).
```
[the remark this answers is not in the document]
Comment (samsp-msft) [reply docs/deployment/azure/application-insights.md:70]: This is difficult as azd does both a provision and deploy action. We have used the term deploy with AZD although that is actually a provision and deploy. I have tried to be clearer now.

File: docs/deployment/azure/application-insights.md
[the remark this answers is not in the document]
Comment (samsp-msft) [reply docs/deployment/azure/application-insights.md:1]: App insights is only provisioned - there is no separate deployment for it- unlike say ACA where you provision the app but can deploy new containers. 

However we call the act of using Azd deployment in the other docs.

**Comment by IEvangelist** [comment]
Hey @samsp-msft - we are missing an image with this path `/deployment/media/app-insights-connection-string.png`. See 
image

**Comment by samsp-msft** [comment]
> Hey @samsp-msft - we are missing an image with this path `/deployment/media/app-insights-connection-string.png`. 

The image is in https://github.com/dotnet/docs-aspire/blob/main/docs/deployment/media/app-insights-connection-string.png and shows for me in the preview.

**Pull request merged**